StylePropertyMap::append is not properly verifying that we don't append CSSVariableReferenceValue style values

pedrovarangot opened the pull request:
#### f2e814363ebf5a9a7902e5cd656ad300053dc129
<pre>
StylePropertyMap::append is not properly verifying that we don&apos;t append CSSVariableReferenceValue style values
<a href="https://rdar.apple.com/147350835">rdar://147350835</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=286767">https://bugs.webkit.org/show_bug.cgi?id=286767</a>

Reviewed by Chris Dumez.

 This fixes the bug by checking that the CSSValue for the style value is
not a variable reference value and throwing an exception if it is. This
commit also modifies an older test that started failing after this change
so that it now passes and still reproduces the crash on versions previous
to its introduction.

* LayoutTests/fast/css/create-columns-onload-crash.html:
* LayoutTests/fast/css/stylepropertymap-append-not-checking-for-variablereferencevalue-expected.txt: Added.
* LayoutTests/fast/css/stylepropertymap-append-not-checking-for-variablereferencevalue.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/append.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/append.tentative.html:
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::append):

Canonical link: <a href="https://commits.webkit.org/292377@main">https://commits.webkit.org/292377@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75aef3f52dc11ae78bc1281cdaaf49a0c99646b3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95674 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100725 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46179 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97718 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15566 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23715 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72962 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30219 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98677 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11659 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86417 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53296 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11372 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4133 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45517 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81554 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4253 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102759 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22728 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16588 "Found 1 new test failure: imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/no-coop-coep.https.any.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82003 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22980 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82434 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81358 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20434 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25947 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3402 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16069 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22696 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27849 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22355 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25831 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24097 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/fast/css/stylepropertymap-append-not-checking-for-variablereferencevalue-expected.txt:
```diff
@@ -0,0 +1,10 @@
+PASS successfullyParsed is true
+
+TEST COMPLETE
+PASS p.attributeStyleMap.append('background-size', body.computedStyleMap().get('-webkit-mask-position')) threw exception TypeError: Values cannot contain a CSSVariableReferenceValue.
```
This is showing after `TEST COMPLETE`, which is evidence that you broke the test. You likely want to be using `testRunner.waitUntilDone()` & `testRunner.notifyDone()` in the test now that it is asynchronous. Otherwise, it WILL be flaky.

pedrovarangot replied:
I am calling `waitUntilDone` at the beginning of the `<script>` tag and `notifyDone` at the end of the test function, what's your suggestion on how to use those?

It passes with `--iterations 1000` on two of my computers and on EWS that I think uses 100 on the -stress runners. Is there any other tool to trigger a failure on a flaky test?

pedrovarangot replied:
Fixed it with `jsTestIsAsync`

rniwa commented on LayoutTests/fast/css/stylepropertymap-append-not-checking-for-variablereferencevalue.html:
```diff
@@ -0,0 +1,25 @@
+<script src="../../resources/js-test.js"></script>
+<script>
+
+window.testRunner?.waitUntilDone();
```
Use `jsTestIsAsync = true` instead.

rniwa commented on LayoutTests/fast/css/stylepropertymap-append-not-checking-for-variablereferencevalue.html:
```diff
@@ -0,0 +1,25 @@
+    body = document.getElementById('body');
+    shouldThrowErrorName("p.attributeStyleMap.append('background-size', body.computedStyleMap().get('-webkit-mask-position'))", "TypeError");
+
+    window.testRunner?.notifyDone();
```
and call `finishJSTest()` here instead.

pedrovarangot replied:
that worked, thanks!

cdumez commented on LayoutTests/fast/css/create-columns-onload-crash.html:
```diff
@@ -2,8 +2,11 @@
+      let cssStyleValue = CSSMathValue.parseAll('columns', 'auto')[0];
+      document.body.attributeStyleMap.append('grid-auto-columns', cssStyleValue);
+    }
+    setTimeout(run, 20);
```
This is a sync test but you made code run asynchronously without adopting `waitUntilDone()`. As commented in my earlier review, this doesn't seem right.

pedrovarangot replied:
Fixed this test too

cdumez commented on LayoutTests/fast/css/create-columns-onload-crash.html:
```diff
@@ -1,9 +1,20 @@
+        let cssStyleValue = CSSMathValue.parseAll('columns', 'auto')[0];
+        document.body.attributeStyleMap.append('grid-auto-columns', cssStyleValue);
+      } finally {
+        if (window.testRunner) {
```
Please do not use curly brackets for one-liners.

cdumez commented on LayoutTests/fast/css/create-columns-onload-crash.html:
```diff
@@ -1,9 +1,20 @@
   onload = () => {
-    let cssStyleValue = CSSMathValue.parseAll('columns', 'auto')[0];
-    document.body.attributeStyleMap.append('grid-auto-columns', cssStyleValue);
+    function run() {
```
Please explain in the commit log WHY you are making this change. It is not obvious.

pedrovarangot replied:
I explained why I made the change but I don't understand why it works.

cdumez commented on LayoutTests/fast/css/stylepropertymap-append-not-checking-for-variablereferencevalue.html:
```diff
@@ -0,0 +1,23 @@
+      }
+    </script>
+  </head>
+  <body id="body" onload="setTimeout(runTest, 20)">
```
Why are we artificially making the test slower here?

pedrovarangot replied:
This test reproduces a crash that we found via fuzzing, when simplifying the fuzzer test which had a lot of other code running I made the test artificially slower to reproduce the crash. Without the timeout the crash doesn't reproduce.

cdumez reviewed (approved):
Please consider line-wrapping your commit log messages. One very long line is painful to read, especially on Github.

Commit:
StylePropertyMap::append is not properly verifying that we don't append CSSVariableReferenceValue style values
rdar://147350835
https://bugs.webkit.org/show_bug.cgi?id=286767

Reviewed by Chris Dumez.

 This fixes the bug by checking that the CSSValue for the style value is
not a variable reference value and throwing an exception if it is. This
commit also modifies an older test that started failing after this change
so that it now passes and still reproduces the crash on versions previous
to its introduction.

* LayoutTests/fast/css/create-columns-onload-crash.html:
* LayoutTests/fast/css/stylepropertymap-append-not-checking-for-variablereferencevalue-expected.txt: Added.
* LayoutTests/fast/css/stylepropertymap-append-not-checking-for-variablereferencevalue.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/append.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/append.tentative.html:
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
(WebCore::StylePropertyMap::append):

Canonical link: https://commits.webkit.org/292377@main
